.github/actions/build_image/action.yml

```
@@ -62,7 +68,42 @@ runs:
         fi
       shell: bash
 
+    - name: Set hash key
+      run: echo "VERSIONKEY=versions-${{ github.ref_name }}=${{ hashFiles(format('images/{0}/stable/**', inputs.image_name)) }}" >> $GITHUB_ENV
+      shell: bash
+
+    - name: Restore last builds versions from cache
+      id: cache
+      uses: actions/cache/restore@v4
+      with:
+        key: ${{ env.VERSIONKEY }}
+        path: versions.freeze
+
+    - name: Check if rebuild is needed
+      run: |
+        # Rebuilds are needed for 
+        # 1. CI is being ran in a PR or is a nightly run
+        # 2. Base images were rebuilt
+        # 3. New pulp versions was released
+        if [[ "${{ github.event_name }}" == "pull_request" || "${{ inputs.image_variant }}" == "nightly" || -n "${{ inputs.built_base_images }}" ]]; then
+          echo "BUILD=true" >> $GITHUB_ENV
+        else
+          if [[ "${{ steps.cache.outputs.cache-hit  }}" == "false" ]]; then
+            echo "BUILD=true" >> $GITHUB_ENV
+          else
+            # Script returns non-zero (100) when new versions are available
+            if python .ci/scripts/check_for_updates.py ${{ github.ref_name }} versions.freeze; then
```

This name would help reading the if clause:
```suggestion
            if python .ci/scripts/check_up_to_date.py ${{ github.ref_name }} versions.freeze; then
```